File: fries-leetcode-training/12-graph/01-number-of-islands.py
"""
LeetCode 200. 岛屿数量
https://leetcode.cn/problems/number-of-islands/

给你一个由'1'（陆地）和'0'（水）组成的二维网格，请你计算网格中岛屿的数量。
岛屿总是被水包围，并且每座岛屿只能由水平方向或竖直方向上相邻的陆地连接形成。

DFS/BFS

时间复杂度：O(m*n)
空间复杂度：O(m*n)
"""

def num_islands(grid):
    """
    岛屿数量 - DFS法
    
    Args:
        grid: 二维网格
        
    Returns:
        岛屿数量
    """
    if not grid or not grid[0]:
        return 0
    
    def dfs(i, j):
        if (i < 0 or i >= len(grid) or 
            j < 0 or j >= len(grid[0]) or 
            grid[i][j] != '1'):
            return
        
        grid[i][j] = '0'  # 标记为已访问
        
        # 四个方向DFS
        dfs(i + 1, j)
        dfs(i - 1, j)
        dfs(i, j + 1)
        dfs(i, j - 1)
    
    count = 0
    for i in range(len(grid)):
        for j in range(len(grid[0])):
            if grid[i][j] == '1':
                dfs(i, j)
                count += 1
    
    return count


def num_islands_bfs(grid):
    """
    岛屿数量 - BFS法
    
    Args:
        grid: 二维网格
        
    Returns:
        岛屿数量
    """
    if not grid or not grid[0]:
        return 0
    
    from collections import deque
    
    directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
    count = 0
    
    for i in range(len(grid)):
        for j in range(len(grid[0])):
            if grid[i][j] == '1':
                queue = deque([(i, j)])
                grid[i][j] = '0'
                
                while queue:
                    x, y = queue.popleft()
                    
                    for dx, dy in directions:
                        nx, ny = x + dx, y + dy
                        if (0 <= nx < len(grid) and 
                            0 <= ny < len(grid[0]) and 
                            grid[nx][ny] == '1'):
                            grid[nx][ny] = '0'
                            queue.append((nx, ny))
                
                count += 1
    
    return count


def test_num_islands():
    """测试函数"""
    # 测试用例1
    grid1 = [
        ["1","1","1","1","0"],
        ["1","1","0","1","0"],
        ["1","1","0","0","0"],
        ["0","0","0","0","0"]
    ]
    result1 = num_islands([row[:] for row in grid1])  # 复制一份避免修改原数组
    print(f"测试1: {result1}")  # 期望: 1
    
    # 测试用例2
    grid2 = [
        ["1","1","0","0","0"],
        ["1","1","0","0","0"],
        ["0","0","1","0","0"],
        ["0","0","0","1","1"]
    ]
    result2 = num_islands([row[:] for row in grid2])
    print(f"测试2: {result2}")  # 期望: 3


if __name__ == "__main__":
    test_num_islands()

File: fries-leetcode-training/12-graph/01-number-of-islands.cpp
/**
 * LeetCode 200. 岛屿数量
 * https://leetcode.cn/problems/number-of-islands/
 * 
 * 给你一个由'1'（陆地）和'0'（水）组成的二维网格，请你计算网格中岛屿的数量。
 * 
 * DFS/BFS
 * 
 * 时间复杂度：O(m*n)
 * 空间复杂度：O(m*n)
 */

class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
        if (grid.empty() || grid[0].empty()) {
            return 0;
        }
        
        int rows = grid.size();
        int cols = grid[0].size();
        int count = 0;
        
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (grid[i][j] == '1') {
                    dfs(grid, i, j);
                    count++;
                }
            }
        }
        
        return count;
    }
    
private:
    void dfs(vector<vector<char>>& grid, int i, int j) {
        if (i < 0 || i >= grid.size() || j < 0 || j >= grid[0].size() || grid[i][j] != '1') {
            return;
        }
        
        grid[i][j] = '0';  // 标记为已访问
        
        // 四个方向DFS
        dfs(grid, i + 1, j);
        dfs(grid, i - 1, j);
        dfs(grid, i, j + 1);
        dfs(grid, i, j - 1);
    }
};

// 测试函数
#include <iostream>
#include <vector>
using namespace std;

void testNumIslands() {
    Solution solution;
    
    vector<vector<char>> grid1 = {
        {'1','1','1','1','0'},
        {'1','1','0','1','0'},
        {'1','1','0','0','0'},
        {'0','0','0','0','0'}
    };
    int result1 = solution.numIslands(grid1);
    cout << "测试1: " << result1 << endl;  // 期望: 1
    
    vector<vector<char>> grid2 = {
        {'1','1','0','0','0'},
        {'1','1','0','0','0'},
        {'0','0','1','0','0'},
        {'0','0','0','1','1'}
    };
    int result2 = solution.numIslands(grid2);
    cout << "测试2: " << result2 << endl;  // 期望: 3
}

int main() {
    testNumIslands();
    return 0;
}
